caffe2/proto/torch.proto

```diff
@@ -4,547 +4,89 @@ import "caffe2/proto/caffe2.proto";
-
-  repeated Dimension stride = 51;
+message ModuleDef {
+  repeated ModuleDef submodules = 1;
+
+  // We suppose to store the modules in one of the following format:
+  //   - methods (static graph or torch script)
+  //   - pickle
+  //   - cpp_arena
+  repeated MethodDef methods = 2;
+  // because the old pickle modules may not be supported by torch_script,
+  // have to stored as pickle_arena at this moment.
+  optional bytes pickle_arena = 3;
+  // should be exposed by the Class Archive, so user can save
+  // module specific data which cannot be store in the graph or torch_script
+  optional bytes cpp_arena = 4;
+
+  // the parameters of this module
+  repeated ParameterDef parameters = 5;
+
+  // the names of inputs and outputs of the module are inferred
+  // from the main method.
 }
 
-// Types
-//
-// The standard ONNX data types.
-message TypeProto {
-
-  message Tensor {
-    // This field MUST NOT have the value of UNDEFINED
-    // This field MUST be present for this version of the IR.
-    optional TensorProto.DataType elem_type = 1;
-    optional TensorShapeProto shape = 2;
-  }
-
-  // Sequence type: List, Tuple
-  message Sequence {
-    // elem_type and elem_type_list cannot appear together.
-    // If all the element types are the same, we use elem_type,
-    // otherwise, we specify the type of each element in elem_type_list.
-    optional TypeProto elem_type = 1;
-    repeated TypeProto elem_type_list = 51;
-    enum SequenceType {
-      UNDEFINED = 0;
-      LIST = 1;
-      TUPLE = 2;
-    }
-    optional SequenceType sequence_type = 52;
-  }
-
-  // Map<K, V>, (not necessary at this moment)
-  message Map {
-    optional TensorProto.DataType key_type = 1;
-    optional TypeProto value_type = 2;
-  }
-
-  // Special type of blobs, based on the type_name, we can choose the right
-  // serializer and deserialzier.
-  message SpecialBlob {
-    optional string type_name = 1;
-  }
-
-  oneof value {
-    // The type of a tensor.
-    Tensor tensor_type = 1;
-    Sequence sequence_type = 4;
-    Map map_type = 5;
-    SpecialBlob special_type = 51;
-  }
-
-  // An optional denotation can be used to denote the whole
-  // type with a standard semantic description as to what is
-  // stored inside. Refer to https://github.com/onnx/onnx/blob/master/docs/TypeDenotation.md#type-denotation-definition
-  // for pre-defined type denotations.
-  optional string denotation = 6;
+enum ProtoVersion {
+  PROTO_VERSION_NEWEST = 0x0000000000000001;
 }
 
-// Operator Sets
-//
-// OperatorSets are uniquely identified by a (domain, opset_version) pair.
-message OperatorSetIdProto {
-  // The domain of the operator set being identified.
-  // The empty string ("") or absence of this field implies the operator
-  // set that is defined as part of the ONNX specification.
-  // This field MUST be present in this version of the IR when referring to any other operator set.
-  optional string domain = 1;
+message ModelDef {
```

Comment:
level 0 is because we want to support both formats, not a real level

levels 1-3 semantically exist in PyTorch programs. In fact, even in MetaNetDef they exist - just the mapping of several nets per "module" is flattened out in one big map.